src/audio/module_adapter/module_adapter.c

```diff
@@ -241,20 +241,33 @@ int module_adapter_prepare(struct comp_dev *dev)
 	}
 
 	/* allocate memory for input buffers */
-	mod->input_buffers = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
-				     sizeof(*mod->input_buffers) * mod->num_input_buffers);
-	if (!mod->input_buffers) {
-		comp_err(dev, "module_adapter_prepare(): failed to allocate input buffers");
-		return -ENOMEM;
+	if (!mod->num_input_buffers) {
+		mod->input_buffers = NULL;
+	} else {
+		mod->input_buffers =
+			rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+				sizeof(*mod->input_buffers) * mod->num_input_buffers);
+
+		if (!mod->input_buffers) {
+			comp_err(dev, "module_adapter_prepare(): failed to allocate input buffers");
+			ret = -ENOMEM;
+			goto in_free;
+		}
 	}
 
 	/* allocate memory for output buffers */
-	mod->output_buffers = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
-				      sizeof(*mod->output_buffers) * mod->num_output_buffers);
-	if (!mod->output_buffers) {
-		comp_err(dev, "module_adapter_prepare(): failed to allocate output buffers");
-		ret = -ENOMEM;
-		goto in_free;
+	if (!mod->num_output_buffers) {
+		mod->output_buffers = NULL;
+	} else {
+		mod->output_buffers =
+			rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+				sizeof(*mod->output_buffers) * mod->num_output_buffers);
+
+		if (!mod->output_buffers) {
+			comp_err(dev, "module_adapter_prepare(): failed to allocate output buffers");
+			ret = -ENOMEM;
+			goto out_free;
```

Comment:
Same here, "out_free" seems wrong, we only need to free the input buffers. This is harmless as output_buffers will be NULL, but the code is still confusing and may lead to bugs later on when other people modify this code and add new cases on top.

Reply:
@kv2019i I rewrote this flow, should be cleaner now.